models/classes/TestSessionService.php

```diff
@@ -287,4 +286,14 @@ public function deleteDeliveryExecutionData(DeliveryExecutionDeleteRequest $requ
 
         return false;
     }
+
+    /**
+     * @param $forReadingOnly
+     * @param string $sessionId
+     * @return bool
+     */
+    private function accessModeChangedToWrite($forReadingOnly, string $sessionId): bool
+    {
+        return !$forReadingOnly && self::$cache[$sessionId][self::SESSION_PROPERTY_SESSION]->isReadOnly();
```

Comment:
I realize the method is private, and if one uses it they probably know what they're doing, but just to be absolutely safe, I would've done something like the following.
```suggestion
        return $this->hasSession($sessionId)
            && !$forReadingOnly
            && self::$cache[$sessionId][self::SESSION_PROPERTY_SESSION]->isReadOnly();
```

Alternatively, introducing a method, getting the session, or loading it if necessary, may be an option as well.

Reply:
At the moment this method is/should be called only if test session exists. But you are right, to be safe it's better to check it as well.